### DISCUSSION TITLE
Is there suport for IKEA air sensor?

### DISCUSSION THREAD

**Discussion by chaveiro** [arendst/Tasmota#12810] [Q&A] [answered]
Thinking in converting the ikea vindriktning  air quality monitor.
Is there tasmota support for the PM1006 Particulate Matter Sensor that comes inside?

See details:

https://community.home-assistant.io/t/ikea-vindriktning-air-quality-sensor/324599/4

https://home-assistant-guide.com/2021/07/20/ikea-makes-an-air-quality-monitor-and-it-has-already-been-hacked/

**Comment by Jason2866** [comment]
No. Since it is serial it should be possible. https://github.com/Hypfer/esp8266-vindriktning-particle-sensor/blob/master/src/SerialCom.h
"Someone" who has the device has to do ;-)

**Reply by linuxrrze** [reply]
I'd love to see support for this device in Tasmota (one image to run all devices would be great!)
I got such a device, and I guess I could do the porting.
However it'd be much easier if I could get some pointers to similar devices implemented in Tasmota and maybe some clues about how to add new devices in general.

**Reply by Jason2866** [reply]
This driver is similar https://github.com/arendst/Tasmota/blob/development/tasmota/xsns_18_pms5003.ino

**Reply by linuxrrze** [reply]
Ok - I'm on it.
However for now when checking PinUsed() I always get _false_ as response.
Am I right, that this function should return _true_ if there's a GPIO configured with this feature?
Any idea why it always returns _false_ - guess I'm missing something here ..

**Reply by barbudor** [reply]
`PinUsed()` should have as parameter the name of a pin and it will return TRUE is such a pin is defined in the Configuration/Template of Tasmota;
So for example for `PinUsed(GPIO_PMS5003_TX)` to be valid, you must have assign `PMS5003_TX` to one GPIO of your Tasmota.

**Reply by barbudor** [reply]
If you are working on a new driver or an update of an existing driver, please do not hesitate to share your GitHub repo so we can easily support you (which may include cloning/downloding your code to test it on our side).
Unfortunately there's no units available close to me, not even for postal delivery :(

**Reply by linuxrrze** [reply]
Hmm... AFAIK that's exactly what I did.
However I copied the PMS50003 .ino file and added a new device to support the IKEA PMS - maybe I missed something elsewhere.

Files I modified:

tasmota/xsns_90_vindriktning.ino (modified copy of PMS5003)
tasmota/support_features.ino (new feature added)
tasmota/tasmota_configurations.h (enabled device build)
tasmota/tasmota_template.h (added sensor name)
tasmota/language/en_GB.h (do get a string in the device selection menu)

For now I can configure a "vindriktning" device on a GPIO port and at least the Init function gets called.

I'll call it a day for now, but my current code can be found here:
https://github.com/linuxrrze/Tasmota/tree/ikea-vindriktning
(and yes: for now it's a dirty hack and will require quite some cleanups once its working ;-) )

**Reply by barbudor** [reply]
Thanks for sharing
There was a mistake with `D_SENSOR_VINDRIKTNING_RX` not located at the right place
```
   D_SENSOR_SDM120_TX "|" D_SENSOR_SDM120_RX "|"
   D_SENSOR_SDM630_TX "|" D_SENSOR_SDM630_RX "|"
-  D_SENSOR_VINDRIKTNING_RX "|"
   D_SENSOR_TM1638_CLK "|" D_SENSOR_TM1638_DIO "|" D_SENSOR_TM1638_STB "|"
   D_SENSOR_DFR562 "|"
```
...
```
   D_SENSOR_INTERRUPT "|"
   D_SENSOR_MCP2515_CS "|"
+  D_SENSOR_VINDRIKTNING_RX "|"
;

```

With this change, I can define the pin and get the correct log
```
00:00:00.063 VindriktningInit
00:00:00.069 VindriktningInit: PinUsed(GPIO_VINDRIKTNING_RX) TRUE
```

I don't yet have the device, may be end of the week. SO I won't be able to do more for now.
I made a PR to your fork

**Reply by linuxrrze** [reply]
Great!
That and reverting some code I played around with gives regular calls to the data collection function.
First real run will have to wait till tomorrow however.
Thanks so much for your help!

**Reply by linuxrrze** [reply]
Ok, looks like data is coming in now.

> stat/tasmota_XXXXXX/STATUS10 {"StatusSNS":{"Time":"2021-08-25T09:09:01","VINDRIKTNING":{"PM1.0":969,"PM2.5":11,"PM10":11}}}
> tele/tasmota_XXXXXX/SENSOR {"Time":"2021-08-25T09:09:04","VINDRIKTNING":{"PM1.0":987,"PM2.5":16,"PM10":16}}
> tasmota/discovery/A848FAXXXXXX/sensors {"sn":{"Time":"2021-08-25T09:09:09","VINDRIKTNING":{"PM1.0":987,"PM2.5":16,"PM10":16}},"ver":1

However I've got still some problems / questions:

Homeassistant only get's the PM2.5 and PM10 right, PM1.0 for some reason has no valid data type (ug/m3) - I cannot see any difference to the other values - so I'm a little lost here.
Maybe it's related to MQTT messages getting sent only once (during startup) - maybe I cleanup up to much of the former code?

**Reply by linuxrrze** [reply]
Looks like I've been a little impatient: MQTT  messages are posted every 5m.
Homeassistant problem persists however.

**Reply by barbudor** [reply]
Command 'teleperiod' allows to change the report.

What is you exact problem with HA?
What are your discovery messages?

**Reply by linuxrrze** [reply]
I guess this is the discovery message you mentioned, right?

> tasmota/discovery/A848FAC037A2/config {"ip":"192.168.1.32","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-C037A2-6050","mac":"A848FAC037A2","md":"Generic","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"9.5.0.7","t":"tasmota_C037A2","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":1,"lt_st":0,"sho":[0,0,0,0],"ver":1}

Problem is this (PM1.0 not recognized as Particle Sensor):
![Screenshot from 2021-08-25 12-14-58](https://user-images.githubusercontent.com/18438286/130772928-b7eaa255-16bf-4d73-999f-4de00f800326.png)

**Reply by barbudor** [reply]
For now, I don't know how home assistant and discovery are working.
I was expecting each driver to fill up something to goes in that discovery message you showed but I can't seem to find that.
I'll dig a little more tonigt.

**Reply by linuxrrze** [reply]
I'll also have a closer look - maybe there's another MQTT message that I missed (and I sent you the wrong one).

**Reply by linuxrrze** [reply]
Looks like autoconfiguration needs to be enabled using "SetOption19 on"

Doing so I get:

> homeassistant/sensor/C037A2_VINDRIKTNING_PM1_0/config {"name":"Tasmota VINDRIKTNING PM1.0","stat_t":"tele/tasmota_C037A2/SENSOR","avty_t":"tele/tasmota_C037A2/LWT","pl_avail":"Online","pl_not_avail":"Offline","uniq_id":"C037A2_VINDRIKTNING_PM1_0","dev":{"ids":["C037A2"]},"unit_of_meas":" ","ic":"mdi:eye","frc_upd":true,"val_tpl":"{{value_json['VINDRIKTNING']['PM1.0']}}"}
> homeassistant/sensor/C037A2_VINDRIKTNING_PM2_5/config {"name":"Tasmota VINDRIKTNING PM2.5","stat_t":"tele/tasmota_C037A2/SENSOR","avty_t":"tele/tasmota_C037A2/LWT","pl_avail":"Online","pl_not_avail":"Offline","uniq_id":"C037A2_VINDRIKTNING_PM2_5","dev":{"ids":["C037A2"]},"unit_of_meas":"µg/m³","ic":"mdi:air-filter","frc_upd":true,"val_tpl":"{{value_json['VINDRIKTNING']['PM2.5']}}"}
> homeassistant/sensor/C037A2_VINDRIKTNING_PM10/config {"name":"Tasmota VINDRIKTNING PM10","stat_t":"tele/tasmota_C037A2/SENSOR","avty_t":"tele/tasmota_C037A2/LWT","pl_avail":"Online","pl_not_avail":"Offline","uniq_id":"C037A2_VINDRIKTNING_PM10","dev":{"ids":["C037A2"]},"unit_of_meas":"µg/m³","ic":"mdi:air-filter","frc_upd":true,"val_tpl":"{{value_json['VINDRIKTNING']['PM10']}}"}

And yes: The first entry is missing the "unit_of_meas" value etc.

**Reply by linuxrrze** [reply]
For better readability:

PM1.0 - not working:

> {
>   "name": "Tasmota VINDRIKTNING PM1.0",
>   "stat_t": "tele/tasmota_C037A2/SENSOR",
>   "avty_t": "tele/tasmota_C037A2/LWT",
>   "pl_avail": "Online",
>   "pl_not_avail": "Offline",
>   "uniq_id": "C037A2_VINDRIKTNING_PM1_0",
>   "dev": {
>     "ids": [
>       "C037A2"
>     ]
>   },
>   "unit_of_meas": " ",
>   "ic": "mdi:eye",
>   "frc_upd": true,
>   "val_tpl": "{{value_json[VINDRIKTNING][PM1.0]}}"
> }

PM 2.5 - working:
> {
>   "name": "Tasmota VINDRIKTNING PM2.5",
>   "stat_t": "tele/tasmota_C037A2/SENSOR",
>   "avty_t": "tele/tasmota_C037A2/LWT",
>   "pl_avail": "Online",
>   "pl_not_avail": "Offline",
>   "uniq_id": "C037A2_VINDRIKTNING_PM2_5",
>   "dev": {
>     "ids": [
>       "C037A2"
>     ]
>   },
>   "unit_of_meas": "µg/m³",
>   "ic": "mdi:air-filter",
>   "frc_upd": true,
>   "val_tpl": "{{value_json[VINDRIKTNING][PM2.5]}}"
> }

**Reply by barbudor** [reply]
You shouldn't need SetOption19 to use the new Tasmota Discovery
SetOption19 enables the old HA discovery which will not be maintained

What you are missing is the `tasmota/discovery/xxx/sensors` message. You don't get it because of a race condition.

This message is build in `TasDiscovery()` by calling `MqttShowSensor()`
But it is done early in the code and you haven't received any valid data yet. So `Vindriktning.valid` is 0 and your `VindriktningShow()` function doesn't return anything valid.

I see 3 possibiliies of solving that:
1- Force `Vindriktning.valid = MIN_INTERVAL_PERIOD;` in `VindriktningInit()` which assume the sensor is there (and will be displayed in the web GUI and in Sensor messages with not valid data) but TasDiscovery() will add it
This is not nice
2- Have a rule which force a new Discovery upon reception of 1st valid value (should be doable using a "Once" rule)
3- find a way in the driver code to trigger TasDiscovery() when .valid change from 0 to non-0

**Reply by barbudor** [reply]
@linuxrrze Discussingfurther with the team and Theo we suggest one of the following:

1) When the driver detect the 1st valid incoming serial frame (Vindriktning.valid changes from zero to non-zero), it sets `TasmotaGlobal.discovery_counter = 1;` which will force TasDicovery() on next 1SEC loop. This time as the data are valid, JSON will be correctly added

2) Instead of not sending any result in Show(), change the logic for:
```
    If not PinUsed(), 
        show nothing
  Else
        if valid
              return JSON with valid values
        else
              return JSON with `null` (without any string quote) instead of value   
```

In 1st case, the discovery message will be (re)sent when the sensor  is discovered for real
In 2nd case, it will always be sent but `null` say it is invalid

**Reply by linuxrrze** [reply]
I tried to implement what you suggested (s. https://github.com/linuxrrze/Tasmota/tree/development).
However the problem persists.

**Reply by barbudor** [reply]
~~You mean the discovery/sensor message doesn't include the PM ? or it doesn't help HA ?~~

I see the discovery topic now:
```
17:21:11.510 MQT: tasmota/discovery/18FE34A6D683/sensors = {"sn":{"Time":"2021-08-25T17:21:11+02:00","ANALOG":{"A0":7},"VINDRIKTNING":{"PM1.0":null,"PM2.5":null,"PM10":null}},"ver":1} (retained)
```
But may be HA doesn't know what to do about a VINDRIKTNING sensor ?

**Reply by linuxrrze** [reply]
That's quite possible, however 2 of the 3 values are accepted as defined - only the first one gets messed up.
I'm just experimenting with a 4th dummy value (hard-coded) to see if it's a first argument thing or something else.

**Reply by linuxrrze** [reply]
Argl... the attribute needs to be called "PM1" instead of "PM1.0" ... works now.

**Reply by linuxrrze** [reply]
I just prepared PR https://github.com/arendst/Tasmota/pull/12976
Feel free to tell me what needs to be changed in order to get merged.

And thanx for the fast and great help! Couldn't have done it without you!

**Reply by barbudor** [reply]
Great!
Should have mine by the week end
👍 😊

**Reply by Senfruecken** [reply]
I try it and have a question.
I use IOBroker and in the sonoff adapter are the values of the "sonoff.0.TestBoard.VINDRIKTNING_PM2" in a separate folder and the objekt is called 5.
"sonoff.0.TestBoard.VINDRIKTNING_PM1" and "sonoff.0.TestBoard.VINDRIKTNING_PM10" not.
Should I change something?

I didn't change anything, I only compiled it online with gitpod.

![Screenshot_20210827-211437_Chrome](https://user-images.githubusercontent.com/59235292/131178796-400b4bdb-2edf-4c17-a5c2-abf178609049.jpg)
![Screenshot_20210827-211431_Chrome](https://user-images.githubusercontent.com/59235292/131178798-8c256e09-120b-4732-a279-8fca489769c8.jpg)
![Screenshot_20210827-211418_Chrome](https://user-images.githubusercontent.com/59235292/131178800-09d5567b-debf-456c-aea9-03f29236f0cc.jpg)

**Reply by barbudor** [reply]
That's a problem with Iobroker which interprets the dot in PM2.5 as a folder level.
I suggest you raise the issue with Iobroker.
Not a Tasmota problem

**Reply by blakadder** [reply]
@linuxrrze did you connect the REST straight to an ESP GPIO?

**Reply by linuxrrze** [reply]
@blakadder Yes (D2). I just followed the instructions as mentioned here: https://github.com/Hypfer/esp8266-vindriktning-particle-sensor

**Reply by Jason2866** [reply]
Seems there is a bug somewhere #13012

**Reply by barbudor** [reply]
Just got one tonight. Some tinkering expected this week

**Reply by Jason2866** [reply]
I found a datasheet for the sensor https://github.com/arendst/Tasmota/issues/13012#issuecomment-909165333
According to that, only PM 2.5 
The orig. hack supports only PM 2.5 too https://github.com/Hypfer/esp8266-vindriktning-particle-sensor/blob/master/src/SerialCom.h#L26-L53

**Reply by chaveiro** [reply]
Just assembled my own today. Great addition.
A remark, the Vindrinking RX pin must have a pullup resistor. On Wemos D1 Mini, it's pin GPIO2.

**Comment by probonopd** [comment, marked as answer]
There is now support for IKEA VINDRIKTNING particle concentration sensor in the development branch:
https://github.com/arendst/Tasmota/pull/12976

Thanks @linuxrrze for making it happen :+1:
